clang/lib/Driver/Driver.cpp

```diff
@@ -6097,10 +6097,29 @@ const char *Driver::GetNamedOutputPath(Compilation &C, const JobAction &JA,
   }
 
   llvm::PrettyStackTraceString CrashInfo("Computing output path");
+
   // Output to a user requested destination?
   if (AtTopLevel && !isa<DsymutilJobAction>(JA) && !isa<VerifyJobAction>(JA)) {
-    if (Arg *FinalOutput = C.getArgs().getLastArg(options::OPT_o))
+    bool IsCLNonPCH =
+        IsCLMode() && !C.getArgs().hasArg(options::OPT__SLASH_Yc) &&
+        (isa<PreprocessJobAction>(JA) || isa<PrecompileJobAction>(JA));
+    bool HasAnyOutputArg = C.getArgs().hasArg(
+        options::OPT_o, options::OPT__SLASH_Fo, options::OPT__SLASH_Fo_COLON);
+
+    Arg *FinalOutput = nullptr;
+    if (IsCLNonPCH && HasAnyOutputArg) {
+      FinalOutput = C.getArgs().getLastArg(
+          options::OPT_o, options::OPT__SLASH_Fo, options::OPT__SLASH_Fo_COLON);
+    } else if (IsCLNonPCH) {
+      FinalOutput = C.getArgs().getLastArg(options::OPT__SLASH_Fo,
+                                           options::OPT__SLASH_Fo_COLON);
+    } else {
+      FinalOutput = C.getArgs().getLastArg(options::OPT_o);
+    }
```

Comment:
I'm having trouble following the logic here. Can you add a comment explaining what's going on? I.e. when is `-o` and/or `/Fo` supposed to be considered?

Reply:
I've updated the OP message with the reasoning for this. This logic in particular is to resolve `lit` test cases that failed when `/Yp` was passed to `clang-cl.exe`, i.e. PCHs were generated. I believe this is because `/Yp` uses `/Fo` for the PCH output, and therefore this invocation needs to use `/o` (or `-o`). 

`HasAnyOutputArg` was written _specifically_ to resolve `llvm-project\clang\test\ClangScanDeps\target-filename.cpp` failing. It feels like a weird hack and I'm open to suggestions.

Reply:
I do hope there's some scope for a response, as I'm not sure how to proceed.

Reply:
You can ping the reviewer @zmodem . The suggested style is no more ping in a week.

Reply:
The logic has to be clear to any reader. "It makes the lit tests pass" is not a good explanation.

I think you need to figure out how these flags are supposed to work together, and then translate that into code.

Reply:
Given [the comment below](https://github.com/llvm/llvm-project/pull/121046#discussion_r1971184880), I believe it is incorrect to pass both `-fmodule-output` and the combination of `--precompile` and `-o` (or `-Fo`). I think there is hence scope to simplify this logic.